File: longest-palindromic-substring/solution5.py
# https://leetcode.com/problems/longest-palindromic-substring/

class Solution5:

    # Two pointer, expand around the center
    # T: O(n^2), S: O(1)
    def longestPalindrome(self, s: str) -> str:
        res_len, res_idx = 0, 0

        for i in range(len(s)):
            # Odd length
            l, r = i, i
            while l >= 0 and r < len(s) and s[l] == s[r]:
                if r - l + 1 > res_len:
                    res_idx = l
                    res_len = r - l + 1
                l -= 1
                r += 1
            # Even length
            l, r = i, i + 1
            while l >= 0 and r < len(s) and s[l] == s[r]:
                if r - l + 1 > res_len:
                    res_idx = l
                    res_len = r - l + 1
                l -= 1
                r += 1

        return s[res_idx:res_idx + res_len]

    # T: O(n^2), S: O(1)
    def longestPalindromeBetterBrute(self, s: str) -> str:
        res = ""
        res_len = 0

        for i in range(len(s)):
            for j in range(i, len(s)):
                l, r = i, j
                while l < r and s[l] == s[r]:
                    l += 1
                    r -= 1

                if l >= r and res_len < (j - i + 1):
                    res = s[i:j + 1]
                    res_len = j - i + 1

        return res

    # T: O(n^2), S: O(n^2)
    def longestPalindromeDp(self, s: str) -> str:
        res_idx, res_len = 0, 0
        n = len(s)

        dp = [[False] * n for _ in range(n)]

        for i in range(n - 1, -1, -1):
            for j in range(i, n):
                if s[i] == s[j] and (j - i <= 2 or dp[i + 1][j - 1]):
                    dp[i][j] = True
                    if res_len < (j - i + 1):
                        res_idx = i
                        res_len = j - i + 1

        return s[res_idx:res_idx + res_len]

    # T: O(n^3), S: O(1)
    def longestPalindromeBrute(self, s: str) -> str:
        res = ""
        res_len = 0
        for i in range(len(s)):
            for j in range(i, len(s)):
                substr = s[i:j + 1]
                if self.isPalindrome(substr) and len(substr) > res_len:
                    res = substr
                    res_len = len(substr)
        return res

    # T: O(k), S: O(1)
    def isPalindrome(self, s: str) -> bool:
        left, right = 0, len(s) - 1
        while left < right:
            if s[left] != s[right]:
                return False
            left += 1
            right -= 1
        return True


def main() -> None:
    sol = Solution5()
    s1 = "babad"
    s2 = "cbbd"
    s3 = "a"
    print(sol.longestPalindrome(s1))            # "bab" or "aba"
    print(sol.longestPalindromeBrute(s1))       # "bab" or "aba"
    print(sol.longestPalindromeBetterBrute(s1))  # "bab" or "aba"
    print(sol.longestPalindrome(s2))            # "bb"
    print(sol.longestPalindromeBrute(s2))       # "bb"
    print(sol.longestPalindromeBetterBrute(s2))  # "bb"
    print(sol.longestPalindrome(s3))            # "a"
    print(sol.longestPalindromeBrute(s3))       # "a"
    print(sol.longestPalindromeBetterBrute(s3))  # "a"


if __name__ == "__main__":
    main()

File: longest-palindromic-substring/Solution5.java
// https://leetcode.com/problems/longest-palindromic-substring/

class Solution5 {

    // Two pointer, expand around the center
    // T: O(n^2), S: O(1)
    public String longestPalindrome(String s) {
        int resLen = 0, resIdx = 0;

        for (int i = 0; i < s.length(); i++) {
            // Odd length
            int l = i, r = i;
            while (l >= 0 && r < s.length() && s.charAt(l) == s.charAt(r)) {
                if (r - l + 1 > resLen) {
                    resIdx = l;
                    resLen = r - l + 1;
                }
                l--;
                r++;
            }
            // Even length
            l = i;
            r = i + 1;
            while (l >= 0 && r < s.length() && s.charAt(l) == s.charAt(r)) {
                if (r - l + 1 > resLen) {
                    resIdx = l;
                    resLen = r - l + 1;
                }
                l--;
                r++;
            }
        }

        return s.substring(resIdx, resIdx + resLen);
    }

    // T: O(n^2), S: O(1)
    public String longestPalindromeBetterBrute(String s) {
        var res = "";
        int resLen = 0;

        for (int i = 0; i < s.length(); i++) {
            for (int j = i; j < s.length(); j++) {
                int l = i, r = j;
                while (l < r && s.charAt(l) == s.charAt(r)) {
                    l++;
                    r--;
                }

                if (l >= r && resLen < (j - i + 1)) {
                    res = s.substring(i, j + 1);
                    resLen = j - i + 1;
                }
            }
        }

        return res;
    }

    // T: O(n^2), S: O(n^2)
    public String longestPalindromeDp(String s) {
        int resIdx = 0, resLen = 0;
        int n = s.length();

        boolean[][] dp = new boolean[n][n];

        for (int i = n - 1; i >= 0; i--) {
            for (int j = i; j < n; j++) {
                if (s.charAt(i) == s.charAt(j) &&
                        (j - i <= 2 || dp[i + 1][j - 1])) {

                    dp[i][j] = true;
                    if (resLen < (j - i + 1)) {
                        resIdx = i;
                        resLen = j - i + 1;
                    }
                }
            }
        }

        return s.substring(resIdx, resIdx + resLen);
    }

    // T: O(n^3), S: O(1)
    public String longestPalindromeBrute(String s) {
        var res = "";
        var resLen = 0;
        for (var i = 0; i < s.length(); i++) {
            for (var j = i; j < s.length(); j++) {
                var substr = s.substring(i, j + 1);
                if (isPalindrome(substr) && substr.length() > resLen) {
                    res = substr;
                    resLen = substr.length();
                }
            }
        }
        return res;
    }

    // T: O(k), S: O(1)
    private boolean isPalindrome(String s) {
        var left = 0;
        var right = s.length() - 1;
        while (left < right) {
            if (s.charAt(left++) != s.charAt(right--)) {
                return false;
            }
        }
        return true;
    }

    public static void main(String[] args) {
        var sol = new Solution5();
        var s1 = "babad";
        var s2 = "cbbd";
        var s3 = "a";
        System.out.println(sol.longestPalindrome(s1)); // "cbbd"
        System.out.println(sol.longestPalindromeBrute(s1)); // "cbbd"
        System.out.println(sol.longestPalindromeBetterBrute(s1)); // "cbbd"
        System.out.println(sol.longestPalindrome(s2)); // "bb"
        System.out.println(sol.longestPalindromeBrute(s2)); // "bb"
        System.out.println(sol.longestPalindromeBetterBrute(s2)); // "bb"
        System.out.println(sol.longestPalindrome(s3)); // "a"
        System.out.println(sol.longestPalindromeBrute(s3)); // "a"
        System.out.println(sol.longestPalindromeBetterBrute(s3)); // "a"
    }

}
